cpp/src/arrow/compare.cc

```diff
@@ -717,6 +719,83 @@ class TypeEqualsVisitor {
   bool result_;
 };
 
+#define VISIT_SIMPLE(SCALAR_TYPE)                                 \
+  Status Visit(const SCALAR_TYPE& left) {                         \
+    const auto& right = checked_cast<const SCALAR_TYPE&>(right_); \
+    result_ = right.value == left.value;                          \
+    return Status::OK();                                          \
+  }
+
+class ScalarEqualsVisitor {
+ public:
+  explicit ScalarEqualsVisitor(const Scalar& right) : right_(right), result_(false) {}
+
+  Status Visit(const NullScalar& left) {
+    result_ = true;
+    return Status::OK();
+  }
+
+  VISIT_SIMPLE(BooleanScalar)
+  VISIT_SIMPLE(Int8Scalar)
+  VISIT_SIMPLE(UInt8Scalar)
+  VISIT_SIMPLE(Int16Scalar)
+  VISIT_SIMPLE(UInt16Scalar)
+  VISIT_SIMPLE(Int32Scalar)
+  VISIT_SIMPLE(UInt32Scalar)
+  VISIT_SIMPLE(Int64Scalar)
+  VISIT_SIMPLE(UInt64Scalar)
+  VISIT_SIMPLE(FloatScalar)
+  VISIT_SIMPLE(DoubleScalar)
+  VISIT_SIMPLE(Date32Scalar)
+  VISIT_SIMPLE(Date64Scalar)
+  VISIT_SIMPLE(Time32Scalar)
+  VISIT_SIMPLE(Time64Scalar)
+  VISIT_SIMPLE(TimestampScalar)
+
+  template <typename T>
+  typename std::enable_if<std::is_base_of<BinaryScalar, T>::value, Status>::type Visit(
+      const T& left_) {
+    const auto& left = checked_cast<const BinaryScalar&>(left_);
+    const auto& right = checked_cast<const BinaryScalar&>(right_);
+    result_ = internal::SharedPtrEquals(left.value, right.value);
```

Comment:
`checked_cast` is unnecessary for upcasts; you can just use `static_cast`
```suggestion
    result_ = internal::SharedPtrEquals(left_.value, static_cast<const T&>(right_).value);
```